src/components/fx/hover.js

```diff
@@ -1993,19 +1993,33 @@ function getCoord(axLetter, winningPoint, fullLayout) {
     var ax = winningPoint[axLetter + 'a'];
     var val = winningPoint[axLetter + 'Val'];
 
+    var trace = winningPoint.trace;
+    var cd = winningPoint.cd;
+    var d = cd[winningPoint.index];
+
     if(ax.type === 'category') val = ax._categoriesMap[val];
     else if(ax.type === 'date') {
-        var period = winningPoint[axLetter + 'Period'];
-        val = ax.d2c(period !== undefined ? period : val);
+        var periodalignment = trace[axLetter + 'periodalignment'];
+        if(periodalignment) {
+            var start = d[axLetter + 'Start'];
+            var end = d[axLetter + 'End'];
+            var diff = end - start;
+            if(periodalignment === 'end') {
+                val += diff;
+            } else if(periodalignment === 'middle') {
+                val += diff / 2;
+            }
+        }
+
+        val = ax.d2c(val);
```

Comment:
@archmoj Can you explain why the previous code was timezone-dependent, and how this fixes it?

Reply:
The previous logic used start and end of period directly which appears to depend on the time zone.
But now we only use the difference between two which should not be impacted by time zone.
We will keep #5861 open until we could further investigate other time zone problems.

Reply:
Ah ok, I think I understand it now. The previous one had the advantage of front-loading more of the calculation, so let's try and get back toward that. I _think_ what's going on is when you pull out the period with:
```js
var period = winningPoint[axLetter + 'Period'];
```
This is already in calcdata format - ie a number - so when we run `ax.d2c` on it we get the legacy convert-from-local-to-UTC behavior. But `val` is in data format - ie a  string - so it needs `ax.d2c`. So the right solution is to just change the one line:
```js
val = ax.d2c(period !== undefined ? period : val);
```
To:
```js
val = period !== undefined ? period : ax.d2c(val);
```

Reply:
Thanks for the hint. There was actually a problem in new code which is fixed now.
Unfortunately  trying `val = period !== undefined ? period : ax.d2c(val);` didn't pass the tests.
However `xStart` and `xEnd` are coming from the calc step so the changes should be fast.

Reply:
Hmm OK - well, this seems to work so let's go with it. Thanks for investigating!